hr_applicant/models/training.py

```diff
@@ -0,0 +1,310 @@
+                    ) + relativedelta(months=rec.course_id.duration, days=-1)
+                end_date = end_date and datetime.datetime.strftime(
+                    end_date, DEFAULT_SERVER_DATE_FORMAT
+                )
+                rec.training_end_date = end_date
+
+    def action_to_be_approve(self):
+        self.write({"state": "to_be_approved"})
+        return True
+
+    def action_approve(self):
+        for rec in self:
+            if not rec.training_attendees:
+                raise ValidationError(
+                    _("Training Attendees should not be Zero!"))
+            rec.write({"state": "approved"})
+        return True
+
+    def action_completed(self):
+        for rec in self:
+            if not rec.attendees_ids:
+                raise ValidationError(
+                    _("You cannot Approve Any Training with no Attendees!")
+                )
+            else:
+                if len(rec.attendees_ids.ids) > rec.training_attendees:
+                    raise ValidationError(
+                        _(
+                            "List of attendees are greater than \
+                           Training Attendees!"
+                        )
+                    )
+                    for attendee in rec.attendees_ids:
+                        if attendee.state not in ("train_completed", "in_complete"):
+                            raise ValidationError(
+                                _(
+                                    "You can not Mark the training as Completed \
+                                   till any of attendee is not in \
+                                   Training Completed or Training incomplete!"
+                                )
+                            )
+        rec.write({"state": "completed"})
+        return True
+
+    def action_cancel(self):
+        for rec in self:
+            for attendee in rec.attendees_ids:
+                if attendee.state not in [
+                    "draft",
+                    "awaiting_training_start",
+                    "in_complete",
+                ]:
+                    raise ValidationError(
+                        _(
+                            "You can not cancel the Training Class if \
+                           all attendees are not in Draft, Awaiting \
+                           Training Start or In complete state!"
+                        )
+                    )
+        self.write({"state": "cancel"})
+        return True
+
+
+class ListOfAttendees(models.Model):
+
+    _name = "list.of.attendees"
+    _rec_name = "class_id"
+    _description = "List Of Attendees"
+
+    @api.constrains(
+        "class_id", "training_start_date", "training_end_date", "date_of_arrival"
+    )
+    def _check_training_dup(self):
+        # self.ensure_one()
+        for rec in self:
+            if str(rec.training_start_date) < datetime.datetime.now().strftime(
+                DEFAULT_SERVER_DATE_FORMAT
+            ):
```

Comment:
Here, we may compare datetime object directly, please check if we can remove string conversion , we may use fields.Datetime.now() rather using python lib datetime.